runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/interceptors/FlexibleChecksumsRequestInterceptor.kt

```diff
@@ -7,102 +7,124 @@ package aws.smithy.kotlin.runtime.http.interceptors
+            return context.protocolRequest
         }
 
-        val headerName = "x-amz-checksum-$checksumAlgorithmName".lowercase()
-        logger.debug { "Resolved checksum header name: $headerName" }
-
-        // remove all checksum headers except for $headerName
-        // this handles the case where a user inputs a precalculated checksum, but it doesn't match the input checksum algorithm
-        req.headers.removeAllChecksumHeadersExcept(headerName)
-
-        val checksumAlgorithm = checksumAlgorithmName?.toHashFunction() ?: throw ClientException("Could not parse checksum algorithm $checksumAlgorithmName")
-
-        if (!checksumAlgorithm.isSupported) {
-            throw ClientException("Checksum algorithm $checksumAlgorithmName is not supported for flexible checksums")
-        }
+        logger.debug { "Calculating checksum using '$checksumAlgorithm'" }
 
-        if (req.body.isEligibleForAwsChunkedStreaming) {
-            req.header("x-amz-trailer", headerName)
+        val request = context.protocolRequest.toBuilder()
 
+        if (request.body.isEligibleForAwsChunkedStreaming) {
             val deferredChecksum = CompletableDeferred<String>(context.executionContext.coroutineContext.job)
 
-            if (req.headers[headerName] != null) {
-                logger.debug { "User supplied a checksum, skipping asynchronous calculation" }
-
-                val checksum = req.headers[headerName]!!
-                req.headers.remove(headerName) // remove the checksum header because it will be sent as a trailing header
-
-                deferredChecksum.complete(checksum)
-            } else {
-                logger.debug { "Calculating checksum asynchronously" }
-                req.body = req.body
-                    .toHashingBody(checksumAlgorithm, req.body.contentLength)
-                    .toCompletingBody(deferredChecksum)
-            }
-
-            req.trailingHeaders.append(headerName, deferredChecksum)
-            return req.build()
+            request.body = request.body
+                .toHashingBody(
+                    checksumAlgorithm,
+                    request.body.contentLength,
+                )
+                .toCompletingBody(
+                    deferredChecksum,
+                )
+
+            request.headers.append("x-amz-trailer", checksumHeader.toString())
+            request.trailingHeaders.append(checksumHeader.toString(), deferredChecksum)
         } else {
-            return super.modifyBeforeSigning(context)
+            checksumAlgorithm.update(
+                request.body.readAll() ?: byteArrayOf(),
+            )
+            request.headers[checksumHeader.toString()] = checksumAlgorithm.digest().encodeBase64String()
         }
+
+        context.executionContext.emitBusinessMetric(checksumAlgorithm.toBusinessMetric())
+        request.headers.removeAllChecksumHeadersExcept(checksumHeader.toString())
```

Comment:
It can be used given two different cases. The first is for when the checksum is provided via a header, we call it, then return. The second time is for when we computed the checksum ourselves, we call it, then return.